File: frontend/src/domain/esg.ts
import { Schema } from "effect"

export const EsgEstimateSchema = Schema.Struct({
  dateUtc: Schema.String, status: Schema.String, dataOrigin: Schema.String, transactionCount: Schema.Number,
  energyLowerWh: Schema.Number, energyBestGuessWh: Schema.Number, energyUpperWh: Schema.Number, emissionsGCo2e: Schema.Number,
  renewableEnergyWh: Schema.Number, nuclearEnergyWh: Schema.Number, fossilEnergyWh: Schema.Number,
})

export const EsgObservationSchema = Schema.Struct({
  observedAtUnixMs: Schema.Number,
  lastProcessedBlock: Schema.Number,
  chainId: Schema.Number,
  contractAddress: Schema.String,
  currentDay: EsgEstimateSchema,
  methodology: Schema.Struct({
    version: Schema.String, annualTransactionsAssumption: Schema.Number,
    lowerEnergyWhPerTransaction: Schema.Number, bestGuessEnergyWhPerTransaction: Schema.Number, upperEnergyWhPerTransaction: Schema.Number,
    emissionsGCo2ePerTransaction: Schema.Number, renewablePercent: Schema.Number,
    nuclearPercent: Schema.Number, fossilPercent: Schema.Number,
    sourceName: Schema.String, sourceUrl: Schema.String, note: Schema.String,
  }),
})
export type EsgObservation = typeof EsgObservationSchema.Type

export const EsgHistorySchema = Schema.Struct({ days: Schema.Array(EsgEstimateSchema), methodology: EsgObservationSchema.fields.methodology })
export type EsgHistory = typeof EsgHistorySchema.Type

File: backend/src/domain/esg.rs
use serde::Serialize;

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EsgMethodology {
    pub version: &'static str,
    pub annual_transactions_assumption: u64,
    pub lower_energy_wh_per_transaction: f64,
    pub best_guess_energy_wh_per_transaction: f64,
    pub upper_energy_wh_per_transaction: f64,
    pub emissions_g_co2e_per_transaction: f64,
    pub renewable_percent: f64,
    pub nuclear_percent: f64,
    pub fossil_percent: f64,
    pub source_name: &'static str,
    pub source_url: &'static str,
    pub note: &'static str,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EsgEstimate {
    pub date_utc: String,
    pub status: &'static str,
    pub transaction_count: u64,
    pub data_origin: &'static str,
    pub energy_lower_wh: f64,
    pub energy_best_guess_wh: f64,
    pub energy_upper_wh: f64,
    pub emissions_g_co2e: f64,
    pub renewable_energy_wh: f64,
    pub nuclear_energy_wh: f64,
    pub fossil_energy_wh: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EsgObservation {
    pub observed_at_unix_ms: u64,
    pub last_processed_block: u64,
    pub chain_id: u64,
    pub contract_address: String,
    pub current_day: EsgEstimate,
    pub methodology: EsgMethodology,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EsgHistory {
    pub days: Vec<EsgEstimate>,
    pub methodology: EsgMethodology,
}
